File: backend/src/routes/analytics.js
import express from 'express';
import authMiddleware from '../middlewares/auth.js';
import { supabase } from '../services/supabaseClient.js';
const router = express.Router();
router.use(authMiddleware);
router.get('/summary', async (req, res, next) => {
    try {
        const userId = req.user.id;
        const expenseResult = await supabase.from('expenses').select('amount').eq('user_id', userId);
        const incomeResult = await supabase.from('income').select('amount').eq('user_id', userId);
        if (expenseResult.error || incomeResult.error) {
            return res.status(400).json({ error: expenseResult.error?.message || incomeResult.error?.message });
        }
        const totalExpense = expenseResult.data?.reduce((acc, row) => acc + Number(row.amount), 0) ?? 0;
        const totalIncome = incomeResult.data?.reduce((acc, row) => acc + Number(row.amount), 0) ?? 0;
        res.json({ totalExpense, totalIncome, savings: totalIncome - totalExpense });
    }
    catch (err) {
        next(err);
    }
});
export default router;

File: backend/routers/analytics.py
from fastapi import APIRouter, Depends
from schemas import User
from dependencies import get_current_user
from mock_db import mock_storage

router = APIRouter()

@router.get("/summary")
def get_analytics_summary(user: User = Depends(get_current_user)):
    user_expenses = [e for e in mock_storage["expenses"] if e.get("user_id") == user.id]
    user_income = [i for i in mock_storage["income"] if i.get("user_id") == user.id]
    
    total_expense = sum([float(e.get("amount", 0)) for e in user_expenses])
    total_income = sum([float(i.get("amount", 0)) for i in user_income])
    
    return {
        "totalExpense": total_expense,
        "totalIncome": total_income,
        "savings": total_income - total_expense
    }
